text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,484 @@
+message. If no "Create" span exists for a message, the context of the related
+"Publish" span must be attached to the message.
+
+### Consumer
+
+For many use cases, it is not possible to rely on the presence of "Process"
+spans for correlating producer with consumer traces: there are cases where a
+dedicated processing operation cannot be identified, or where processing
+happens in a different trace. Furthermore, processing operations often are not
+covered by messaging libraries and SDKs, but take place in application code.
+Consistently creating spans for "Processing" operations would require either
+effort from the application owner to correctly instrument those operations, or
+additional capabilities of messaging libraries and SDKs (e. g. hooks for
+processing callbacks, which can then be instrumented by the libraries or SDKs).
+
+While it is possible to create "Process" spans and correlate those with
+consumer traces in certain cases, this is not something that can be generally
+required. Therefore, it is more feasible to require the creation of "Deliver"
+spans (for push-based APIs) or "Receive" spans (for pull-based APIs) to
+correlate producer with consumer traces.
+
+#### Instrumenting push-based scenarios
+
+In push-based consumer scenarios, the delivery of messages is not initiated by
+the application code. Instead, callbacks or handlers are registered and then
+called by messaging SDKs to forward messages to the application.
+
+A "Deliver" span covers the call of such a callback or handler and should link
+to the "Create" spans of all messages that are forwarded via the respective
+call. Depending on the use case, "Deliver" spans can correlate with "Process"
+spans or other spans modelling processing operations.
+
+#### Instrumenting pull-based scenarios
+
+In pull-based consumer scenarios, the delivery of messages is requested by the
+application code. This usually involves a blocking call, which returns zero or
+more messages on completion.
+
+A "Receive" span covers such calls and should link to the "Create" spans of all
+messages that are forwarded via the respective call. Depending on the use case,
+"Receive" spans can correlate with "Process" spans or other spans modelling
+processing operations.
+
+#### General considerations for both push-based and pull-based scenarios
+
+The operations modelled by "Deliver" or "Receive" spans do not strictly refer
+to receiving the message from intermediaries, but instead refer to the
+application receiving messages for processing. If messages are fetched from the
+intermediary and forwarded to the application in one go, the whole operation
+might be covered by a "Deliver" or "Receive" span. However, libraries or SDKs
+might pre-fetch messages from intermediaries and cache those messages, and only
+forward messages to the application at a later time. In this case, the
+operation of pre-fetching and caching should not be covered by the "Deliver" or
+"Receive" spans.
```

Comment:
I don't get what it actually means. Is the pre-fetching an implementation detail and an SDK optimization but in the end a message is anyway delivered or received? I mean when the SDK pre-fetches the messages and put them into a cache, no spans are created but when they are provided to the application, the "Deliver" or "Receiver" will be created. Is my assumption correct?

Reply:
> I mean when the SDK pre-fetches the messages and put them into a cache, no spans are created but when they are provided to the application, the "Deliver" or "Receiver" will be created. Is my assumption correct?

Yes, mostly.

SDKs can surely create spans that cover implementation specific details, but we don't intend to cover this with v1 of these conventions. This might be covered by transport-level instrumentation (e. g. AMQP instrumentation).